packages/analytics-js/src/components/configManager/ConfigManager.ts

```diff
@@ -1,21 +1,150 @@
 import { state } from '@rudderstack/analytics-js/state';
+import { Destination } from '@rudderstack/analytics-js/state/types';
+import { resolveDataPlaneUrl } from './util/dataPlaneResolver';
+import { getIntegrationsCDNPath } from './util/cdnPaths';
+import { getSDKUrlInfo } from './util/commonUtil';
+import { IConfigManager, SourceConfigResponse } from './types';
+import { filterEnabledDestination } from './util/filterDestinations';
 
-class ConfigManager {
+class ConfigManager implements IConfigManager {
+  httpClient: IHttpClient;
+  errorHandler?: IErrorHandler;
+  logger?: ILogger;
+  hasErrorHandler = false;
+  hasLogger = false;
+
+  constructor(httpClient: IHttpClient, errorHandler?: IErrorHandler, logger?: ILogger) {
+    this.errorHandler = errorHandler;
+    this.logger = logger;
+    this.httpClient = httpClient;
+    this.hasErrorHandler = Boolean(this.errorHandler);
+    this.hasLogger = Boolean(this.logger);
+  }
+
+  /**
+   * A function to validate, construct and store loadOption, lifecycle, source and destination
+   * config related information in global state
+   */
   init() {
-    // TODO: get writekey, dataplaneUrl and loadOptions from state to use
-    this.getConfig();
+    validateLoadArgs(state.lifecycle.writeKey.value, state.lifecycle.dataPlaneUrl.value);
+    // determine the sourceConfig url
+    const intgCdnUrl = getIntegrationsCDNPath(
+      'process.package_version',
+      state.loadOptions.value.lockIntegrationsVersion as boolean,
+      state.loadOptions.value.destSDKBaseURL,
+    );
+
+    // determine if the staging SDK is being used
+    // TODO: deprecate this in new version and stop adding '-staging' in filenames
+    const { isStaging } = getSDKUrlInfo();
+
+    // set the final load option in state
+    batch(() => {
+      // set application lifecycle state in state
+      if (state.loadOptions.value.logLevel) {
+        state.lifecycle.logLevel.value = state.loadOptions.value.logLevel;
+      }
+      state.lifecycle.integrationsCDNPath.value = intgCdnUrl;
+      if (state.loadOptions.value.configUrl) {
+        // TODO: pass the load option for lockedIntegrations version as query param in the future for metrics capturing
+        state.lifecycle.sourceConfigUrl.value = `${state.loadOptions.value.configUrl}/sourceConfig/?p=process.module_type&v=process.package_version&writeKey=${state.lifecycle.writeKey.value}`;
+      }
+      state.lifecycle.isStaging.value = isStaging;
+    });
+    this.fetchSourceConfig();
   }
 
-  getConfig() {
-    // Once done call all ready
-    this.onReady();
+  /**
+   * A callback function that is executed once we fetch the source config response.
+   * Use to construct and store information that are dependent on the sourceConfig.
+   * @param res source config response
+   */
+  processResponse(res: SourceConfigResponse | string | undefined) {
+    if (!res || typeof res === 'string') {
```

Comment:
if it shoudl always be object better to check that if type is not object instead of string

Reply:
Done